File: src/main/kotlin/tables/Templates.kt
package tables

import org.jetbrains.exposed.v1.core.Table
import org.jetbrains.exposed.v1.datetime.datetime

object Templates : Table() {
    val id = integer("id").autoIncrement()
    val name = varchar("name", 255)
    val description = text("description").nullable()
    val imageFileId = reference("image_file_id", StorageFiles.id).nullable()
    val categoryId = integer("category_id").references(TemplateCategories.id)
    val order = integer("order")
    val preSuspensionCategoryId = integer("pre_suspension_category_id").references(TemplateCategories.id).nullable()
    val createdAt = datetime("created_at")
    val updatedAt = datetime("updated_at")

    override val primaryKey = PrimaryKey(id)
}

File: db/schema/templates.ts
import {
    pgTable,
    serial,
    varchar,
    text,
    integer,
    timestamp,
    bigint,
    pgEnum,
    uniqueIndex,
} from "drizzle-orm/pg-core";

export const templatecategorySpecialTypeEnum = pgEnum("CategorySpecialType", [
    "Main",
    "Suspension",
]);

export const templateCategories = pgTable(
    "TemplateCategory",
    {
        id: serial("id").primaryKey(),
        name: varchar("name", { length: 255 }).notNull(),
        description: text("description"),
        parentCategoryId: integer("parentCategoryId"),
        order: integer("order"),
        specialType: templatecategorySpecialTypeEnum("specialType"),
        createdAt: timestamp("createdAt", { precision: 3 }).notNull(),
        updatedAt: timestamp("updatedAt", { precision: 3 }).notNull(),
    },
    (table) => [
        uniqueIndex("TemplateCategory_specialType_key").on(
            table.specialType,
        ),
    ],
);

export const templates = pgTable("Template", {
    id: serial("id").primaryKey(),
    name: varchar("name", { length: 255 }).notNull(),
    description: text("description"),
    imageFileId: bigint("imageFileId", { mode: "bigint" }),
    categoryId: integer("categoryId").notNull(),
    order: integer("order").notNull(),
    preSuspensionCategoryId: integer("preSuspensionCategoryId"),
    createdAt: timestamp("createdAt", { precision: 3 }).notNull(),
    updatedAt: timestamp("updatedAt", { precision: 3 }).notNull(),
});

export const templateConfigKeyEnum = pgEnum("TemplateConfigKey", [
    "MAX_BACKGROUND_SIZE",
    "ALLOWED_FILE_TYPES",
]);

export const templateConfigs = pgTable("TemplateConfigs", {
    key: templateConfigKeyEnum("key").primaryKey(),
    value: text("value").notNull(),
});